Gifski/util.swift

```diff
@@ -1865,3 +1865,187 @@ extension BinaryFloatingPoint {
 		return (self * divisor).rounded(rule) / divisor
 	}
 }
+
+/// Since errors in Swift can be any type, there's no easy way to store context info, so we use this struct for that and pass it around.
+struct ErrorContext {
+	/**
+	Stores all the error contexts.
+
+	We're using a `NSMutableDictionary` here as we can't use associated objects with `Error` as it's a generic protocol. We can't use `NSMapTable`, which would be perfect for this, as it requires the type AnyObject. Storing it like this means we retain the error forever, but it should not really have much impact on memory usage.
+	*/
+	static let store = NSMutableDictionary()
+
+	static func get(from error: Error) -> ErrorContext? {
+		return store[error] as? ErrorContext
+	}
+
+	let originalError: Error
+	let function: String
+	let file: String
+	let line: Int
+	let column: Int
+	let stack: String
+
+	init(
+		error originalError: Error,
+		function: String = #function,
+		file: String = #file,
+		line: Int = #line,
+		column: Int = #column
+	) {
+		self.originalError = originalError
+		self.function = function
+		self.file = file.nsString.deletingPathExtension.nsString.lastPathComponent
+		self.line = line
+		self.column = column
+		/// TODO: Need to find a way to demangle the symbols
+		self.stack = Thread.callStackSymbols.joined(separator: "\n")
+	}
+}
+
+/// Wrapper error to store more information about the original error
+struct SoftError {
+	private let message: String
+
+	/// Create a new soft error
+	init(
+		_ message: String,
+		function: String = #function,
+		file: String = #file,
+		line: Int = #line,
+		column: Int = #column
+	) {
+		self.message = message
+
+		ErrorContext.store[self] = ErrorContext(
+			error: self,
+			function: function,
+			file: file,
+			line: line,
+			column: column
+		)
+	}
+
+	/// Wrap an existing error to give it more context
+	init(
+		_ error: Error,
+		function: String = #function,
+		file: String = #file,
+		line: Int = #line,
+		column: Int = #column
+	) {
+		// Prevent double-wrapping
+		if let error = error as? SoftError {
+			self = error
+			return
+		}
+
+		self.message = error.localizedDescription
+
+		ErrorContext.store[self] = ErrorContext(
+			error: error,
+			function: function,
+			file: file,
+			line: line,
+			column: column
+		)
+	}
+}
+
+extension SoftError: LocalizedError {
+	var errorDescription: String? {
+		return message
+	}
+}
+
+extension SoftError: GithubReportableError {}
+
+protocol GithubReportableError: RecoverableError {}
+
+extension GithubReportableError {
```

Comment:
I am not sure whether this is the best solution. I guess many people use this app who don't even have a Github account. Also, we don't really want hundreds of duplicates which manually need to be closed.
Reporting the error to a custom backend and just comparing some stuff before automatically posting it on Github would be a better option in my opinion. This way users will report it more often and the Github issues don't become a mess...